src/mlpack/core/math/make_alias.hpp

```diff
@@ -19,52 +19,66 @@ namespace mlpack {
-               const size_t numRows,
-               const size_t numCols,
+template<typename InVecType, typename OutVecType>
+void MakeAlias(OutVecType& v,
+               const InVecType& oldVec,
+               const size_t numElems,
+               const size_t offset = 0,
                const bool strict = true,
-               const typename std::enable_if_t<!IsCube<MatType>::value>* = 0)
+               const typename std::enable_if_t<IsVector<InVecType>::value>* = 0)
 {
   // We use placement new to reinitialize the object, since the copy and move
   // assignment operators in Armadillo will end up copying memory instead of
   // making an alias.
-  m.~MatType();
-  new (&m) MatType(newMem, numRows, numCols, false, strict);
+  typename InVecType::elem_type* newMem =
+    const_cast<typename InVecType::elem_type*>(oldVec.memptr()) + offset;
+  v.~OutVecType();
+  new (&v) OutVecType(newMem, numElems, false, strict);
 }
 
 /**
- * Reconstruct `c` as an alias around the memory` newMem`, with size `numRows` x
- * `numCols` x `numSlices`.
+ * Reconstruct `m` as an alias around the memory `newMem`, with size `numRows` x
+ * `numCols`.
  */
-template<typename CubeType>
-void MakeAlias(CubeType& c,
-               typename CubeType::elem_type* newMem,
+template<typename InMatType, typename OutMatType>
+void MakeAlias(OutMatType& m,
+               const InMatType& oldMat,
                const size_t numRows,
                const size_t numCols,
-               const size_t numSlices,
+               const size_t offset = 0,
                const bool strict = true,
-               const typename std::enable_if_t<IsCube<CubeType>::value>* = 0)
+               const typename std::enable_if_t<IsMatrix<InMatType>::value>* = 0)
 {
   // We use placement new to reinitialize the object, since the copy and move
   // assignment operators in Armadillo will end up copying memory instead of
   // making an alias.
-  c.~CubeType();
-  new (&c) CubeType(newMem, numRows, numCols, numSlices, false, strict);
+  typename InMatType::elem_type* newMem =
+    const_cast<typename InMatType::elem_type*>(oldMat.memptr()) + offset;
+  m.~OutMatType();
+  new (&m) OutMatType(newMem, numRows, numCols, false, strict);
 }
 
 /**
- * Make `m` an alias of `in`, using the given size.
+ * Reconstruct `c` as an alias around the memory` newMem`, with size `numRows` x
+ * `numCols` x `numSlices`.
  */
-template<typename eT>
-void MakeAlias(arma::Mat<eT>& m,
-               const arma::Mat<eT>& in,
+template<typename InCubeType, typename OutCubeType>
+void MakeAlias(OutCubeType& c,
+               const InCubeType& oldCube,
                const size_t numRows,
                const size_t numCols,
-               const bool strict = true)
+               const size_t numSlices,
+               const size_t offset = 0,
+               const bool strict = true,
+               const typename std::enable_if_t<IsCube<InCubeType>::value>* = 0)
```

Comment:
Might want an `enable_if` for all of these to ensure that `eT` is the same for `InCubeType` and `OutCubeType`... for that matter, you could just keep the parameters as `arma::Cube<eT>`.  That might avoid some danger of wrong template type inference.  This won't be a problem for Bandicoot, once we add a `MakeAlias` version for `coot::Mat<eT>`.

Reply:
Oh, I now see, reading later, that we have some cases where we want to make a `Cube` that is an alias of a `Mat`.  So I guess this does need to stay as a template parameter.  But shouldn't we do SFINAE on `OutCubeType` instead of `InCubeType`?  e.g. `c` should always be a `Cube` here, but it doesn't matter if `oldCube` is actually a cube, since we are just taking its memory location.